Use Jss instance counter  in class names generator #644

kof opened the pull request.

oliviertassinari commented on src/Jss.js:
```diff
@@ -27,7 +27,11 @@ declare var __VERSION__: string
+let instanceCounter = 0
+
 export default class Jss {
+  id = instanceCounter++
```
Is it equivalent to having it in the constructor?

kof replied:
yes, its new property initializers syntax in es-something

oliviertassinari commented on src/utils/createGenerateClassName.js:
```diff
@@ -33,12 +28,18 @@ export default (): generateClassName => {
     }
 
-    return `${prefix + rule.key}-${jssCounter}-${ruleCounter}`
+    return `${prefix + rule.key}-${moduleId}${jssId && `-${jssId}`}-${ruleCounter}`
```
Can `jssId` be falsy?

kof replied:
excluded in the line above

oliviertassinari commented on src/utils/createGenerateClassName.js:
```diff
@@ -2,12 +2,7 @@
-if (global[ns] == null) global[ns] = 0
-// In case we have more than one JSS version.
-const jssCounter = global[ns]++
+import moduleId from './moduleId'
```
What's the use case for the moduleId? Isn't jssId replacing it?

kof replied:
no, we need both, added a better comment about it https://github.com/cssinjs/jss/pull/648/commits/88ffe5e688a61e633a3a70ebc47257a89f15bca9

oliviertassinari commented on src/utils/moduleId.js:
```diff
@@ -3,5 +3,8 @@ import global from './global'
 if (global[ns] == null) global[ns] = 0
 
-// In case we have more than one module version.
+// Bundle may contain multiple JSS versions at the same time. In order to identify
```
I still don't get it. How is the bundle containing multiple JSS versions on issue? Doesn't the issue coming from the different JSS instance?

kof replied:
If bundle contains multiple JSS package versions, modules are evaluated separately, which means JSS instance counter will start from 0 for each version.

kof replied:
I could replace the module id with a JSS instance counter which always starts using a global counter, like I do with moduleId right now, so there will be no need for module id. I thought this implementation will be easier to reason about.

oliviertassinari commented:
I came to realize the implication of this PR. It means that people will be able to have multiple version of JSS in their bundle. Effectively, shipping two times the same large CSS-in-JSS runtime to their users. Shouldn't we discourage it by not making it work and by having jss as a peer dependency instead?

kof commented:
We had support for multiple JSS versions before that as well.

Some time ago we introduced rules counter reset through createGenerateClassName option and this allowed user to have 2 Jss instances with the rules counter starting at 0 and clashing (same JSS module).

So now we just fixed this by adding also Jss instance counter, so that user can have many Jss instances and reset the rules counter whenever they want without clashes.


That being said, we don't have other choice rather than support multiple versions, because user might need this in some situation for migration  purposes. We could add a warning though for the case user is not aware, but we should not forbid this. 

I am happy to add a warning for dev.
